File: app/src/main/java/com/spbstu/sneakerhunter/ShopsCardAdapter.java
package com.spbstu.sneakerhunter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class ShopsCardAdapter extends RecyclerView.Adapter<ShopsCardAdapter.ViewHolder> {

    private ShopsCardAdapter.Listener listener;
    //TODO replace list of elements with new shop class and refactor onBindViewHolder using new list
    private  List<Element> elements;

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private CardView cardView;

        public ViewHolder(CardView cv){
            super(cv);
            cardView = cv;
        }
    }

    interface Listener {
        void onClick(int position);
    }

    ShopsCardAdapter(List<Element> elements) {
        this.elements = new ArrayList<>();
        this.elements = elements;
    }

    @NonNull
    @Override
    public ShopsCardAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        CardView cardView = (CardView) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.shops_cardview, parent, false);

        return new ShopsCardAdapter.ViewHolder(cardView);
    }

    public void setListener(ShopsCardAdapter.Listener listener) {
        this.listener = listener;
    }

    @Override
    public void onBindViewHolder(@NonNull ShopsCardAdapter.ViewHolder holder, final int position) {
        CardView cardView = holder.cardView;

        TextView shopNameTextView = (TextView) cardView.findViewById(R.id.shop_name);
        shopNameTextView.setText(elements.get(position).getName());

        TextView elementPriceTextView = (TextView) cardView.findViewById(R.id.element_price);
        elementPriceTextView.setText(elements.get(position).getPriceString());

        ImageButton elementDeletionButton = (ImageButton) cardView.findViewById(R.id.element_deletion);


        cardView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    listener.onClick(position);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return elements.size();
    }
}

File: app/src/main/java/com/spbstu/sneakerhunterkotlin/adapters/ShopsCardAdapter.kt
package com.spbstu.sneakerhunterkotlin.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.spbstu.sneakerhunterkotlin.R
import com.spbstu.sneakerhunterkotlin.server_list.Sneaker
import java.util.*


class ShopsCardAdapter internal constructor(elementsParam: List<Sneaker>) :
    RecyclerView.Adapter<ShopsCardAdapter.ViewHolder>() {
    private var listener: Listener? = null

    //TODO replace list of elements with new shop class and refactor onBindViewHolder using new list
    private val elements: List<Sneaker> = elementsParam

    class ViewHolder(val cardView: CardView) : RecyclerView.ViewHolder(cardView)

    interface Listener {
        fun onClick(position: Int)
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        val cardView = LayoutInflater.from(parent.context)
            .inflate(R.layout.shops_cardview, parent, false) as CardView
        return ViewHolder(cardView)
    }

    fun setListener(listener: Listener?) {
        this.listener = listener
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) {
        val cardView = holder.cardView
        val shopNameTextView =
            cardView.findViewById<View>(R.id.shop_name) as TextView
        shopNameTextView.setText(elements[position].name)
        val elementPriceTextView =
            cardView.findViewById<View>(R.id.element_price) as TextView
        elementPriceTextView.setText(elements[position].money)
        val elementDeletionButton =
            cardView.findViewById<View>(R.id.element_deletion) as ImageButton
        cardView.setOnClickListener {
            if (listener != null) {
                listener!!.onClick(position)
            }
        }
    }

    override fun getItemCount(): Int {
        return elements.size
    }

}